File: frontend/src/services/addresses.ts
import { apiClient } from './apiClient'
import type { Address, AddressInput } from '@/types/address'
import type { Envelope } from '@/types/catalog'

export async function listAddresses(): Promise<Address[]> {
  const { data } = await apiClient.get<Envelope<Address[]>>('/addresses')
  return data.data
}

export async function createAddress(payload: AddressInput): Promise<Address> {
  const { data } = await apiClient.post<Envelope<Address>>('/addresses', payload)
  return data.data
}

export async function updateAddress(id: number, payload: AddressInput): Promise<Address> {
  const { data } = await apiClient.patch<Envelope<Address>>(`/addresses/${id}`, payload)
  return data.data
}

export async function deleteAddress(id: number): Promise<void> {
  await apiClient.delete(`/addresses/${id}`)
}

File: backend/app/api/v1/account/addresses.py
from fastapi import APIRouter, Depends, Response
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_db
from app.models.auth import User
from app.schemas.address import AddressIn, AddressOut
from app.services.address_service import AddressService

router = APIRouter(prefix="/addresses", tags=["addresses"])


@router.get("", response_model=list[AddressOut])
async def list_addresses(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await AddressService(db).list_addresses(user.id)


@router.post("", response_model=AddressOut, status_code=201)
async def create_address(
    payload: AddressIn, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)
):
    return await AddressService(db).create_address(user.id, payload)


@router.patch("/{address_id}", response_model=AddressOut)
async def update_address(
    address_id: int,
    payload: AddressIn,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return await AddressService(db).update_address(user.id, address_id, payload)


@router.delete("/{address_id}", status_code=204)
async def delete_address(
    address_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)
):
    await AddressService(db).delete_address(user.id, address_id)
    return Response(status_code=204)
